clients/src/main/java/org/apache/kafka/common/network/KafkaChannel.java

```diff
@@ -108,6 +116,19 @@ public boolean isMute() {
         return muted;
     }
 
+    public boolean isInMutableState() {
+        //some requests (like disconnect) may not require memory allocation.
+        //do we do not want to just mute channels where otherwise the next request received
+        //could be immediately executed regardless of memory.
+        //a channel would mute itself if it cannot allocate memory.
+        //also, we cannot mute channels where any memory required has already been allocated.
+        //(for requests that dont require memory receive.memoryAllocated() would also be true
+        if (receive == null || receive.memoryAllocated())
+            return false;
+        //also cannot mute if underlying transport is unmutable right now
+        return transportLayer.isInMutableState();
```

Comment:
@radai-rosenblatt It is also about which layers need to know about these different aspects. Does `SslTransportLayer` really need to know about mutability of buffers? And the reason I suggested the change was because `KafkaChannel.isInMutableState()` should return false if either of the conditions in `this.ready()` is false (i.e. transport layer handshake or authenticating). I don't think it makes sense for transport later or authenticator to have to worry about mutability of buffers.